pallets/nomad/da-bridge/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+/// Edit this file to define custom logic or remove it if it is not needed.
+/// Learn more about FRAME and the core library of Substrate FRAME pallets:
+/// <https://docs.substrate.io/v3/runtime/frame>
+pub use pallet::*;
+
+mod message;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use da_primitives::traits::ExtendedHeader;
+	use frame_support::{
+		pallet_prelude::*,
+		sp_runtime::{
+			traits::{CheckedSub, Header},
+			ArithmeticError,
+		},
+	};
+	use frame_system::{ensure_signed, pallet_prelude::OriginFor};
+	use home::Pallet as Home;
+	use nomad_core::TypedMessage;
+	use primitive_types::H256;
+
+	use crate::message::{DABridgeMessages, DataRootMessage};
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + home::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		#[pallet::constant]
+		type DABridgePalletId: Get<H256>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// Block number to block hash mapping
+	#[pallet::storage]
+	#[pallet::getter(fn finalized_block_number_to_block_hash)]
+	pub type FinalizedBlockNumberToBlockHash<T: Config> =
+		StorageMap<_, Twox64Concat, T::BlockNumber, T::Hash>;
+
+	// Genesis config
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {}
+
+	#[cfg(feature = "std")]
+	impl Default for GenesisConfig {
+		fn default() -> Self { Self {} }
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig {
+		fn build(&self) {}
+	}
+
+	// Hooks
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_finalize(block_number: T::BlockNumber) {
+			let last = block_number.checked_sub(&1u32.into()).unwrap();
+
+			// Store every finalized block's corresponding hash
+			let hash = frame_system::Pallet::<T>::block_hash(last);
+			FinalizedBlockNumberToBlockHash::<T>::insert(last, hash);
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		DataRootDispatched {
+			destination_domain: u32,
+			recipient_address: H256,
+			block_number: T::BlockNumber,
+			data_root: H256,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InitializationError,
+		BlockNotFinal,
+		HashOfBlockNotMatchBlockNumber,
+		DABridgeMessageExceedsMaxMessageSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		[u8; 32]: From<T::AccountId>,
+		H256: From<T::Hash>,
+		u32: From<T::BlockNumber>,
+	{
+		/// Dispatch a data root message to the home if the header is valid.
+		#[pallet::weight(100)]
+		pub fn try_dispatch_data_root(
+			origin: OriginFor<T>,
+			destination_domain: u32,
+			recipient_address: H256,
+			header: T::Header,
```

Comment:
Can make this change for `do_dispatch_data_root`. But we do need the full header for `try_dispatch_data_root` because we need to validate the full header by checking block hash is in finalized mapping